### ISSUE TITLE
what to do when a view gets longer than the screen as it flickers badly on updates

### ISSUE THREAD

**Issue by KyleAMathews** [vadimdemedes/ink#359]
![ink-flickering](https://user-images.githubusercontent.com/71047/89714383-f63d9c80-d952-11ea-843b-4268be14cfe2.gif)

Is there any easy way to detect when this is happening so we can take corrective measures?

**Comment by vadimdemedes** [comment]
Hm, not really, this is up to your terminal to render the text, so the flickering is on them. For example, while iTerm may flicker, built-in terminal app in macOS almost never does.

As for the code here, what is the UI supposed to do? I will probably be able to help better if I'm more familiar with your use case.

**Comment by H3RSKO** [comment]
We noticed in our app (gitgud) that iTerm2 would cause it to become almost unusable due to flickering. If you switch terminals to hyper or the default mac terminal it will almost completely fix the issue. I think it is just an iTerm2 bug.

**Comment by mAAdhaTTah** [comment]
FWIW, I feel like lately iTerm2 has been having issues like this that are definitely unrelated to Ink. Just streaming significant of output in general has been weird. Maybe the most recent update had a bug or something?

**Comment by marcus13371337** [comment]
Hello,

I'm having real issues running my application in "fullscreen" on iTerm. The screen is flickering really badly when there are a lot of updates to the screen. It seems as iTerm queues up all the frames and shows them one by one instead of just showing the latest frame.

My application is a circle-ci dashboard. I was inspired by k9s (https://github.com/derailed/k9s) which runs really smoothly on iTerm (written in go though). Have a look at my project to see if you can find any gotchas:
https://github.com/marcus13371337/circle-dash

**Comment by H3RSKO** [comment]
It seems to be an iTerm issue. Try using a different terminal.

**Comment by marcus13371337** [comment]
Yes, it's definitely an iTerm issue, however, if I wanted to distribute my cli I want it to run smoothly in every terminal. Since there are applications running smoothly on iTerm (k9s as an example), I guess there are some hacks/features ink can implement in order to make the iTerm experience better

**Comment by marcus13371337** [comment]
Think I found a solution to the problem which I've presented in the following PR:
https://github.com/vadimdemedes/ink/pull/413

@H3RSKO Would you be super kind to try it with gitgud in iTerm and confirm that it fixes the flickering, would be super nice. It's not 100% bug-free though

**Comment by pie6k** [comment]
![CleanShot 2021-02-28 at 12 47 01](https://user-images.githubusercontent.com/7311462/109417248-143c1e80-79c3-11eb-9ef6-2c8b11556ea1.gif)

For me this happens in mac default terminal as well (using oh-my-zsh) as soon as I set main container Box height to `process.stdout.rows` and update any part of the content (happens in iTerm as well)

**Comment by royshouvik** [comment]
The changes in the linked PR drastically improves the flickering in iTerm for me. Although I needed to make sure to make the box height 1 less than `process.stdout.rows`.

**Comment by p00dle** [comment]
Just chiming in to say that I had the same problem on Windows, and @royshouvik 's solution worked for me as well for Windows Terminal both cmd and PowerShell. Thank you for the tip!

**Comment by executionreverted** [comment]
up. im making a chat app, whenever i type in input, even tho its memozied and using local states, it feels like whole app re-renders and having flickering on iterm2 and warp.
macos and kitty terminal dont have this issue. any idea how to fix it ?

**Comment by pkkim** [comment]
Here's a tip for people trying to reduce flicker. If your app meets these conditions:
1. Most of the content that appears in the terminal is static, and the content that can change at any given time is shorter in height than the terminal.
2. Importantly, that static content is _entirely above_ all the dynamic content. If this isn't true, your layout may break.

In this case, you can use \<Static\> to eliminate re-renders of anything that doesn't change and is above the dynamic content. As long as the dynamic content is shorter than terminal height, the flickering should be minimal or nonexistent. The only caveat is that _any_ dynamic content will render below the static content.

This seems obvious in hindsight to me, but it's only slightly hinted at in the [docs on Static](https://github.com/vadimdemedes/ink?tab=readme-ov-file#static). The example in the README is helpful.

One trick we used in our use case is that we had a scrolling list of items of which only the last one could change. So we put an array consisting of all but the last item into Static, and made sure the last one was relatively short.

**Comment by technillogue** [comment]
I believe your underlying problem is using ansi-escapes terrible clearTerminal implementation (^[2J^[3J^[H, erase screen and erase full scrollback and move to home) instead of something like ^[H^[0J (move to home, erase to bottom of screen). you should use alternate screen mode like every other well-behaved TUI that doesn't eat the entire scrollback, or just count the number of lines that need to be erased to redraw.

You also need to use cursorSavePosition before rendering and cursorRestorePosition after rendering instead of just leaving the cursor at the bottom of the screen

**Comment by sindresorhus** [comment]
Ink now has incremental rendering (only redraws changed lines) and Synchronized Update Mode (terminals batch the update atomically), both of which directly address the flickering. These were added in recent versions. Closing as resolved.

If you still encounter flickers, please open a new issue with some reproducible example.
